src/ds1302.c

```diff
@@ -2,130 +2,122 @@
 // http://datasheets.maximintegrated.com/en/ds/DS1302.pdf
 //
 
-#pragma callee_saves ds_writebit,ds_readbit
+#pragma callee_saves sendbyte,readbyte
 #pragma callee_saves ds_writebyte,ds_readbyte
 
 #include "ds1302.h"
 
 #define MAGIC_HI  0x5A
 #define MAGIC_LO  0xA5
 
-
-void ds_ram_config_init(uint8_t config[4]) {
+void ds_ram_config_init() {
     uint8_t i;
     // check magic bytes to see if ram has been written before
     if ( (ds_readbyte( DS_CMD_RAM >> 1 | 0x00) != MAGIC_LO || ds_readbyte( DS_CMD_RAM >> 1 | 0x01) != MAGIC_HI) ) {
         // if not, must init ram config to defaults
         ds_writebyte( DS_CMD_RAM >> 1 | 0x00, MAGIC_LO);
         ds_writebyte( DS_CMD_RAM >> 1 | 0x01, MAGIC_HI);
 
-        //for (i=0; i<4; i++)
-            //ds_writebyte( DS_CMD_RAM >> 1 | (i+2), 0x00);
-	ds_ram_config_write(config);
+	ds_ram_config_write();
 	return;
     }
     
     // read ram config
-    for (i=0; i<4; i++)
-        config[i] = ds_readbyte(DS_CMD_RAM >> 1 | (i+2));
+    for (i=0; i!=4; i++)
+        config_table[i] = ds_readbyte(DS_CMD_RAM >> 1 | (i+2));
 }
 
-void ds_ram_config_write(uint8_t config[4]) {
+void ds_ram_config_write() {
     uint8_t i;
-    for (i=0; i<4; i++)
-        ds_writebyte( DS_CMD_RAM >> 1 | (i+2), config[i]);
-}
-
-void ds_writebit(__bit b) {
-    _nop_; _nop_;
-    DS_IO = b;
-    DS_SCLK = 1;
-    _nop_; _nop_;
-    DS_SCLK = 0;
-}
-
-__bit ds_readbit() {
-    __bit b;
-    _nop_; _nop_;
-    b = DS_IO;
-    DS_SCLK = 1;
-    _nop_; _nop_;
-    DS_SCLK = 0;
-    return b;
+    for (i=0; i!=4; i++)
+        ds_writebyte( DS_CMD_RAM >> 1 | (i+2), config_table[i]);
+}
+
+void sendbyte(uint8_t b)
+{
+  b;
+  __asm
+	push	ar7
+        mov     a,dpl
+	mov	r7,#8
+00001$:
+	nop
+	nop
+        rrc     a
+        mov     _P1_1,c
```

Comment:
Jens,
It is possible to put symbols in the inline assembly, but for variable access a "_" must appear in the assembly output, thus the _P1_1 instead of P1_1 as defined by DS_SCLK.
The workaround is quite easy, I just figured it out : I have defined _DS_SCLK as _P1_1 (same for DS_IO & DS_CE), and used _DS_SCLK in the __asm/__endasm block. It produces the same output, but lisibility is better in the .c source.